File: Everything-app-backend/everything-app/src/main/java/com/Finn/everything_app/model/StudyNote.java
package com.Finn.everything_app.model;

import jakarta.persistence.*;
import lombok.Data;
import java.time.LocalDateTime;


@Entity
@Table(name = "study_notes")
@Data
public class StudyNote {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, length = 200)
    private String title;

    @Column(nullable = false, columnDefinition = "TEXT")
    private String content;

    // Die alte, gefälschte Hierarchie: hier stand eine Ordner-ID, die serverseitig nie
    // existierte. Die Spalte bleibt stehen (ddl-auto=update nimmt nichts weg), wird aber nicht
    // mehr beschrieben — der Baum hängt jetzt an parent.
    @Column(length = 100)
    private String category;

    @Column(length = 500)
    private String tags;

    @Column(name = "is_favorite")
    private Boolean isFavorite = false;

    @Column(name = "created_at")
    private LocalDateTime createdAt;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @Column(name = "last_reviewed_at")
    private LocalDateTime lastReviewedAt;


    private String filePath;
    private String fileType;


    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "course_id")
    private Course course;

    /**
     * Die Elternseite; {@code null} heißt Wurzel.
     *
     * Kein {@code isFolder}, kein Typ-Enum: jeder Knoten ist eine Seite, eine Seite mit Kindern
     * bekommt in der UI ein Aufklapp-Dreieck. Genau die Ordner/Notiz-Dualität hatte den
     * category-Hack erzwungen.
     *
     * Bewusst OHNE {@code cascade}: das Löschen eines Teilbaums läuft explizit über den Service
     * (siehe StudyNoteService.deleteNote), damit die Kurszähler stimmen.
     */
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "parent_id")
    private StudyNote parent;

    /** Position unter den Geschwistern. */
    @Column(name = "order_index")
    private Integer orderIndex = 0;

    /** Ein Emoji vor dem Titel. Acht Zeichen, weil ein Emoji mehrere Codepoints haben kann. */
    @Column(length = 8)
    private String icon;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
        normalizeDefaults();
    }

    /**
     * Bestandszeilen haben nach ddl-auto=update NULL in order_index — der Feldinitialisierer
     * greift nur bei frisch konstruierten Objekten, nicht bei geladenen. Ohne das fliegt beim
     * ersten Sortieren eine NPE beim Unboxing.
     */
    @PostLoad
    protected void normalizeDefaults() {
        if (orderIndex == null) orderIndex = 0;
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }


}

File: Everything-app-frontend/everything_app/lib/models/study_note.dart

class StudyNote {
  final int? id;
  final String title;
  final String content;
  final int? courseId;
  final String? courseName;
  final String? category;
  final String? tags;
  final bool isFavorite;

  /// Seitenbaum: `null` heisst Wurzelseite. Es gibt keinen Ordner-Typ — jeder Knoten ist eine
  /// Seite, eine Seite mit Kindern bekommt ein Aufklapp-Dreieck.
  final int? parentId;
  final int orderIndex;
  final String? icon;

  final DateTime? createdAt;
  final DateTime? updatedAt;
  final DateTime? lastReviewedAt;

  StudyNote({
    this.id,
    required this.title,
    required this.content,
    this.courseId,
    this.courseName,
    this.category,
    this.tags,
    this.isFavorite = false,
    this.parentId,
    this.orderIndex = 0,
    this.icon,
    this.createdAt,
    this.updatedAt,
    this.lastReviewedAt,
  });

  // JSON zu StudyNote
  factory StudyNote.fromJson(Map<String, dynamic> json) {
    return StudyNote(
      id: json['id'],
      title: json['title'],
      content: json['content'],
      courseId: json['courseId'],
      courseName: json['courseName'],
      category: json['category'],
      tags: json['tags'],
      isFavorite: json['isFavorite'] ?? false,
      parentId: json['parentId'],
      orderIndex: json['orderIndex'] ?? 0,
      icon: json['icon'],
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'])
          : null,
      updatedAt: json['updatedAt'] != null
          ? DateTime.parse(json['updatedAt'])
          : null,
      lastReviewedAt: json['lastReviewedAt'] != null
          ? DateTime.parse(json['lastReviewedAt'])
          : null,
    );
  }

  /// Ohne parentId und orderIndex: die Baumstruktur gehoert dem Server und aendert sich nur
  /// ueber /move und /reorder. Beim ANLEGEN wird parentId separat mitgegeben.
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'content': content,
      'courseId': courseId,
      'category': category,
      'tags': tags,
      'icon': icon,
      'isFavorite': isFavorite,
    };
  }

  // Copy with
  StudyNote copyWith({
    int? id,
    String? title,
    String? content,
    int? courseId,
    String? courseName,
    String? category,
    String? tags,
    bool? isFavorite,
    int? parentId,
    int? orderIndex,
    String? icon,
    DateTime? createdAt,
    DateTime? updatedAt,
    DateTime? lastReviewedAt,
  }) {
    return StudyNote(
      id: id ?? this.id,
      title: title ?? this.title,
      content: content ?? this.content,
      courseId: courseId ?? this.courseId,
      courseName: courseName ?? this.courseName,
      category: category ?? this.category,
      tags: tags ?? this.tags,
      isFavorite: isFavorite ?? this.isFavorite,
      parentId: parentId ?? this.parentId,
      orderIndex: orderIndex ?? this.orderIndex,
      icon: icon ?? this.icon,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      lastReviewedAt: lastReviewedAt ?? this.lastReviewedAt,
    );
  }

  // Parse Tags als Liste
  List<String> get tagList {
    if (tags == null || tags!.isEmpty) return [];
    return tags!.split(',').map((t) => t.trim()).toList();
  }

  @override
  String toString() => 'StudyNote(id: $id, title: $title)';
}